# This file was generated by Rcpp::compileAttributes
# Generator token: 10BE3573-1514-4C36-9D1C-5A225CD40393

ipf <- function(seed, row_margins, col_margins, iter = 1000L, tol = 1e-10) {
    .Call('methsim_ipf', PACKAGE = 'methsim', seed, row_margins, col_margins, iter, tol)
}

#' Sample H.
#'
#' Sample a column of H for each row of H with the probability of sampling
#' column j for row i given by H[i, j].
#'
#' @param H a numeric matrix where H[i, j] is the probability of sampling
#' column j for row i, i.e., rowMeans(H) == 1.
#'
#' @keywords internal
#'
#' @return an integer vector of length equal to nrow(H), where each element is
#' the column sampled for that row.
.sampleH <- function(H) {
    .Call('methsim_sampleH', PACKAGE = 'methsim', H)
}

#' Sample Z.
#'
#' Simulate sequencing reads by sampling a subset of rows from a given column
#' of H for each read.
#'
#' @param Z an integer matrix where H[i, j] is the methylation state of the
#' i-th methylation locus on the j-th haplotype.
#' @param h an integer vector of length equal to the number of reads. Each
#' element is the haplotype from which each read is to be sampled, i.e.,
#' Z[, j].
#' @param fh an integer vector of length equal to the number of reads. Each
#' element is the position of the first methylation loci ("first hit") from
#' which read is to be sampled, i.e., Z[i, ].
#' @param cqh an integer vector of length equal to the number of reads. Each
#' element is the number of methylation loci that each read overlaps.
#'
#' @keywords internal
#'
#' @return a list of length equal to the number of simulated reads. Each list
#' element is the sequenced/sampled methylation state for that read.
.sampleZ <- function(Z, h, fh, cqh) {
    .Call('methsim_sampleZ', PACKAGE = 'methsim', Z, h, fh, cqh)
}

#' Simulate sequencing errors.
#'
#' Simulate sequencing errors by modifying \code{z} in place.
#'
#' @param z an integer vector of sampled methylation states.
#' @param u an integer vector with the same length as \code{z} of random
#' variables distributed as Uniform(0, 1).
#' @param errorRate the (combined) probability of an "error", where an error
#' may be sequencing error or error in the bisulfite-conversion process.
#'
#' @note \strong{WARNING}: \code{z} is modified \strong{in place}.
#'
#' @return \code{NULL}. \code{z} is modified \strong{in place}.
#'
#' @keywords internal
.simErrorInPlace <- function(z, u, errorRate) {
    invisible(.Call('methsim_simErrorInPlace', PACKAGE = 'methsim', z, u, errorRate))
}

#' Simulate a single "haplotype" of a methylome (Z).
#'
#' @param beta_by_region the beta-value (average methylation level) for each
#' methylation locus in the genome.
#' @param lor_by_pair the within-fragment co-methylation between each pair of
#' methylation loci in the genome. Should be log odds-ratios using base-2
#' logarithms. The length of this should be equal to the number of methylation
#' loci in the genome minus the number of chromosomes (seqnames).
#' @param seqnames_one_tuples the chromosome (seqname) of each methylation
#' locus in the genome, i.e., \code{seqnames(one_tuples)}.
#' @param u a vector of Uniform(0, 1) random variables used in choosing the
#' next state of the process. Must have same length as \code{beta_by_region}.
#'
#' @note Random number generation is performed outside of this function (at
#' the R level in a single thread) in order to simplify this function. If
#' the vector of Uniform(0, 1) random variables ('u') is to be simulated
#' within this function at the C++ level then great care must be taken,
#' especially if this function is subsequently called in parallel.
#'
#' @return an integer vector of simulated methylation states for each
#' methylation locus in the genome; 0 = unmethylated and 1 = methylated.
.simulateZ <- function(beta_by_region, lor_by_pair, seqnames_one_tuples, u) {
    .Call('methsim_simulateZ', PACKAGE = 'methsim', beta_by_region, lor_by_pair, seqnames_one_tuples, u)
}

# Register entry points for exported C++ functions
methods::setLoadAction(function(ns) {
    .Call('methsim_RcppExport_registerCCallable', PACKAGE = 'methsim')
})
